PhysicsTools/NanoAOD/plugins/DSAMuonTableProducer.cc

```diff
@@ -0,0 +1,440 @@
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "TrackingTools/IPTools/interface/IPTools.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/VertexReco/interface/VertexFwd.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+class DSAMuonTableProducer : public edm::global::EDProducer<> {
+public:
+  DSAMuonTableProducer(const edm::ParameterSet& iConfig)
+      : dsaMuonTag_(consumes<std::vector<reco::Track>>(iConfig.getParameter<edm::InputTag>("dsaMuons"))),
+        muonTag_(consumes<std::vector<pat::Muon>>(iConfig.getParameter<edm::InputTag>("muons"))),
+        vtxTag_(consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("primaryVertex"))),
+        bsTag_(consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("beamspot"))),
+        transientTrackBuilderToken_(esConsumes(edm::ESInputTag("", "TransientTrackBuilder"))) {
+    produces<nanoaod::FlatTable>("DSAMuon");
+  }
+
+  ~DSAMuonTableProducer() override {}
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("dsaMuons")->setComment("input displaced standalone muon collection");
+    desc.add<edm::InputTag>("muons")->setComment("input muon collection");
+    desc.add<edm::InputTag>("primaryVertex")->setComment("input primary vertex collection");
+    desc.add<edm::InputTag>("beamspot")->setComment("input beamspot collection");
+    descriptions.add("DSAMuonTable", desc);
+  }
+
+private:
+  void produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const override;
+
+  bool passesDisplacedID(const reco::Track& dsaMuon) const;
+  int getMatches(const pat::Muon& muon, const reco::Track& dsaMuon, float minPositionDiff = 1e-6) const;
+  int getDTMatches(const pat::Muon& muon, const reco::Track& dsaMuon, float minPositionDiff = 1e-6) const;
+  int getCSCMatches(const pat::Muon& muon, const reco::Track& dsaMuon, float minPositionDiff = 1e-6) const;
+  int getTotSegments(const reco::Track& dsaMuon) const;
+  int getDTSegments(const reco::Track& dsaMuon) const;
+  int getCSCSegments(const reco::Track& dsaMuon) const;
+
+  edm::EDGetTokenT<std::vector<reco::Track>> dsaMuonTag_;
+  edm::EDGetTokenT<std::vector<pat::Muon>> muonTag_;
+  edm::EDGetTokenT<reco::VertexCollection> vtxTag_;
+  edm::EDGetTokenT<reco::BeamSpot> bsTag_;
+  edm::ESGetToken<TransientTrackBuilder, TransientTrackRecord> transientTrackBuilderToken_;
+};
+
+void DSAMuonTableProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  float minPositionDiffForMatching = 1e-6;
+
+  edm::Handle<std::vector<reco::Track>> dsaMuonHandle;
+  iEvent.getByToken(dsaMuonTag_, dsaMuonHandle);
+  edm::Handle<std::vector<pat::Muon>> muonHandle;
+  iEvent.getByToken(muonTag_, muonHandle);
```

Comment:
Seems like these could be
```suggestion
  const std::vector<reco::Track>& dsaMuon = iEvent.get(dsaMuonTag_);
  const std::vector<pat::Muon>& muonHandle = iEvent.get(muonTag_);
```
same for the other `Handle`+`getByToken()` pairs (while they work, this suggestion would be less code and tiny bit faster by avoiding the checks in `Handle` de-references).